README.md
# OAuth 2.0 with Keycloak and Clojure (WIP!)

![Authorization Code Flow image](images/authorization_code_flow.png)

The purpose of this project is to showcase the use of Keycloak as an authorization server. We will be using Clojure for building the authorization client and other protected resources.
As time goes on, other features will be added and documented. The possible features can be seen in the listing below:

* Protecting resources with JWT tokens and added security with PKCE
* Redis for caching
* UI with Clojurescript (Hiccup, Reagent & maybe Re-frame as well ;))
* CI/CD (not sure about the technologies yet...)
* Tests

## Keycloak

Keycloak is used as our authorization server and access management solution.

I'll be running a containerized keycloak, using the docker-compose tool.  
To boot the container, run the command:
```
$ docker-compose -f keycloak-postgres.yml up
```
Keycloak will be available at this address: `http://localhost:8080/auth`  
You can log in as administrator with these credentials: `admin / admin`   
The following command will allow us to import the preconfigured keycloak testing environment:
```
$ docker exec -it mykeycloak /opt/jboss/keycloak/bin/standalone.sh \
-Djboss.socket.binding.port-offset=100 \
-Dkeycloak.migration.action=import \
-Dkeycloak.migration.provider=singleFile \
-Dkeycloak.migration.file=/opt/jboss/keycloak/imports/my_realm.json
```
The following elements will be generated for us automatically: 
* Test realm -> `Sso-test`
* Authorization client -> `billingApp`
* Token checker client -> `tokenChecker` 
* Test user with credentials -> `bob / return0`
* New client scope -> `getBillingService`   
The `getBillingService` scope is required in the token when requesting services from the protected resource server. 

If you choose to create your own test keycloak environment, use the following configuration for your client:
* Client protocol -> `openid-connect`
* Access type -> `confidential`
* Root URL -> `http://localhost:3000`

To list the Keycloak endpoints in use in our authorization client, load your test realm page and click on the endpoints link. You should get a similar listing as in the pic below:

![Openid-configuration image](images/openid-configuration.png)

Lastly add a new user for login purposes.

## Authorization Client (Clojure)
The Clojure projects are managed with the Clojure CLI tool. The `deps.edn` file holds the configuration and the needed dependencies.  
To run the service, cd into the `clj-auth-service` directory and execute the command:
```
$ clj -M -m core.sso-clojure
```
The authorization service will be running on port `3000`.

The `Services` link on the landing page is mapped to a handler, that fetches data from the `Billing` service which is protected.   
We need to be logged in and have a valid token with the right scope in order to successfully request the services.

## Billing Service Client (Clojure)
To run the service, CD into the `billing-service` directory and run the command:
```
$ clj -M -m core.billing-service
```
The protected resource server will be running on port `4000`.

## License

* [Apache License, Version 2.0](https://www.apache.org/licenses/LICENSE-2.0)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-01-01
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-01-01
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/core/sso-clojure/compare/0.1.1...HEAD
[0.1.1]: https://github.com/core/sso-clojure/compare/0.1.0...0.1.1


clj-auth-service/test/core/sso_clojure_test.clj
(ns core.sso-clojure-test
  (:require [clojure.test :refer :all]
            [core.sso-clojure :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


billing-service/test/core/billing_service_test.clj
(ns core.billing-service-test
  (:require [clojure.test :refer :all]
            [core.billing-service :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


billing-service/src/core/billing_service.clj
(ns core.billing-service
  (:require  [cheshire.core :refer [generate-string parse-string decode]]
             [clj-http.client :as client]
             [clojure.walk :refer [keywordize-keys]]
             [muuntaja.middleware :as muuntaja]
             [reitit.ring :as reitit]
             [ring.adapter.jetty :as jetty]
             [ring.util.http-response :as response]
             [ring.util.response :refer [redirect]]
             [ring.middleware.reload :refer [wrap-reload]]
             [ring.middleware.params :refer [wrap-params]]
             [ring.util.codec :as codec]
             [selmer.parser :as selmer]
             [selmer.middleware :refer [wrap-error-page]]
             [taoensso.timbre :as log])
  (:import [org.apache.commons.codec.binary Base64]))


(def config {:token-introspection "http://localhost:8080/auth/realms/sso-test/protocol/openid-connect/token/introspect"
             :client-id "tokenChecker"
             :client-password "b5393f3f-9149-4d6f-bf7d-edc36a504388"})

(def services {:services ["electric" "phone" "internet" "water"]})

(defn wrap-nocache [handler]
  (fn [request]
    (-> request
        handler
        (assoc-in [:headers "Pragma"] "no-cache"))))

(defn wrap-formats [handler]
  (-> handler
      (muuntaja/wrap-format)))

(defn response-handler [request]
  (response/ok
   (str "<html><body>your IP is: "
        (:remote-addr request)
        "</body></html>")))

(defn- get-token [request]
  (cond
    ;; headers
    (some? (get-in request [:headers "authorization"])) 
    (do 
      (log/info ::get-token {:method :authorization})
      (get-in request [:headers "authorization"]))
    
    ;; form-params
    (some? (get-in request [:form-params "access_token"]))
    (do
      (log/info ::get-token {:method :form-body})
      (get-in request [:form-params "access_token"]))
    
    ;; query
    (some? (get-in request [:query-params "access_token"]))
    (do 
      (log/info ::get-token {:method :query-params})
      (get-in request [:query-params "access_token"]))
    
    :else nil))

(defn- validate-token [token]
  (let [response
        (client/post (:token-introspection config)
                     {:headers {"Content-Type" "application/x-www-form-urlencoded"}
                      :basic-auth [(:client-id config) (:client-password config)]
                      :form-params {:token_type_hint "requesting_party_token"
                                    :token (clojure.string/replace token #"Bearer " "")}})
        active (-> response
                   :body
                   parse-string
                   (get "active"))]
    (log/info ::validate-token {:active active})
    active))

(defn- decode-token-claim [token]
  (let [claim (-> token
                  (clojure.string/split #"\.")
                  second
                  Base64/decodeBase64
                  String.
                  decode)]
    (log/info ::check-token-scope {:claim claim})
    claim))

(defn services-handler [request]
  (log/info ::services-handler {:headers (keys (:headers request))})
  (if-let [token (get-token request)]
    (do
      (let [valid (validate-token token)
            scope (-> (decode-token-claim token)
                      (get "scope")
                      (clojure.string/split #" ")
                      set)]
        (log/info ::services-handler {:valid-token valid :scope scope})
        (cond
          (false? valid) 
          (do 
            (log/info ::services-handler {:error "Invalid token"})               
            (response/bad-request "Invalid token"))
          (nil? (get scope "getBillingService")) 
          (do
            (log/info ::services-handler {:error "Invalid token scope. Required scope [getBillingService]"})
            (response/bad-request "Invalid scope"))
          :else (response/ok (generate-string services)))))
    (do
      (log/info ::services-handler {:error "Missing access token"})
      (response/forbidden "Missing access token"))))

(defn test-request-handler [request]
  (let [{params :params 
         form-params :form-params 
         query-params :query-params
         headers :headers} request]
    (log/info ::services-post {:request {:form-params form-params
                                         :query-params query-params
                                         :params params
                                         :headers headers
                                         :result (validate-token (get-token request))}})
    (response/ok "OK")))

(def routes
  [["/billing/v1" {:middleware [wrap-params]}
    ["/services" {:get services-handler
                  :post test-request-handler}]]])

(def handler
  (reitit/routes
   (reitit/ring-handler
    (reitit/router routes))
   (reitit/create-resource-handler
    {:path "/"})
   (reitit/create-default-handler
    {:not-found
     (constantly (response/not-found "404 - Page not found"))
     :method-not-allowed
     (constantly (response/method-not-allowed "405 - Not allowed"))
     :not-acceptable
     (constantly (response/not-acceptable "406 - Not acceptable"))})))

(defn -main []
  (jetty/run-jetty
   (-> #'handler
       wrap-nocache
       wrap-reload)
   {:port 4000
    :join? false}))


clj-auth-service/src/core/sso_clojure.clj
(ns core.sso-clojure
  (:require
   [cheshire.core :refer [generate-string parse-string]]
   [clj-http.client :as client]
   [clojure.walk :refer [keywordize-keys]]
   [muuntaja.middleware :as muuntaja]
   [reitit.ring :as reitit]
   [ring.adapter.jetty :as jetty]
   [ring.util.http-response :as response]
   [ring.util.response :refer [redirect]]
   [ring.middleware.reload :refer [wrap-reload]]
   [ring.util.codec :as codec]
   [selmer.parser :as selmer]
   [selmer.middleware :refer [wrap-error-page]]
   [slingshot.slingshot :refer [try+ throw+]]
   [taoensso.timbre :as log]))

(def config
  {:auth-url "http://localhost:8080/auth/realms/sso-test/protocol/openid-connect/auth"
   :logout-url "http://localhost:8080/auth/realms/sso-test/protocol/openid-connect/logout"
   :token-endpoint "http://localhost:8080/auth/realms/sso-test/protocol/openid-connect/token"
   :client-id "billingApp"
   :client-password "fe0a7e01-8b66-4706-8f37-e7d333c29e6f"
   :redirect-uri "http://localhost:3000/auth-code-redirect"
   :landing-page "http://localhost:3000"
   :services-endpoint "http://localhost:4000/billing/v1/services"})

(def app-var (atom {}))

(defn wrap-nocache [handler]
  (fn [request]
    (-> request
        handler
        (assoc-in [:headers "Pragma"] "no-cache"))))

(defn wrap-formats [handler]
  (-> handler
      (muuntaja/wrap-format)))

(defn response-handler [request]
  (response/ok
   (str "<html><body>your IP is: "
        (:remote-addr request)
        "</body></html>")))

(defn home-handler [request]
  (response/ok
   (selmer/render-file "login.html" {:title "~=[λ RuL3z!]=~"
                                     :session (:session-state @app-var)
                                     :code (:code @app-var)
                                     :access-token (get-in @app-var [:token :access_token])
                                     :refresh-token (get-in @app-var [:token :refresh_token])
                                     :scope (get-in @app-var [:token :scope])
                                     :services (:services @app-var)})))

(defn login-handler [request]
  ;; create a redirect URL for authentication endpoint.
  (let [client_id (:client-id config)
        redirect_uri (:redirect-uri config)
        query-string (client/generate-query-string
                      {:client_id client_id
                       :response_type "code"
                       :redirect_uri redirect_uri})
        auth-url (:auth-url config)]
    (redirect (str auth-url "?" query-string))))

(defn get-token []
  (client/post (:token-endpoint config)
               {:headers {"Content-Type" "application/x-www-form-urlencoded"}
                :basic-auth [(:client-id config) (:client-password config)]
                :form-params {:grant_type "authorization_code"
                              :code (:code @app-var)
                              :redirect_uri (:redirect-uri config)
                              :client_id (:client-id config)}}))

(defn- exchange-token []
  (let [token (get-token)]
    (swap! app-var assoc :token (-> (:body token) parse-string keywordize-keys))
    (log/info ::exchange-token {:token (get-in @app-var [:token :token_type])})
    (redirect (:landing-page config))))

(defn auth-code-redirect [request]
  (log/info ::auth-redirect {:query-string (:query-string request)})
  (let [query-params (-> request :query-string codec/form-decode keywordize-keys)
        landing-page (:landing-page config)]
    (swap! app-var assoc :code (:code query-params)
           :session-state (:session_state query-params))
    (exchange-token)))

(defn logout-handler [request]
  (let [query-string (client/generate-query-string {:redirect_uri (:landing-page config)})
        logout-url (str (:logout-url config) "?" query-string)]
    (reset! app-var {})
    (redirect logout-url)))



(defn services-handler [request]
  (if-let [services (try+ 
                     (client/get (:services-endpoint config)
                                 {:headers {"Authorization" (str "Bearer " (get-in @app-var [:token :access_token]))}
                                  :socket-timeout 500
                                  :connection-timeout 500})
                     (catch Object _
                       (log/error ::services "Upstream error")
                       nil))]
    (do
      (swap! app-var assoc :services (-> services :body parse-string (get "services")))
      (log/info ::services {:services services})
      (redirect (:landing-page config)))
    (do 
      (swap! app-var assoc :service {:services []})
      (redirect (:landing-page config)))))

(def routes
  [["/" {:get home-handler}]
   ["/login" {:get login-handler}]
   ["/services" {:get services-handler}]
   ["/logout" {:get logout-handler}]
   ["/auth-code-redirect" {:get auth-code-redirect}]])

(def handler
  (reitit/routes
   (reitit/ring-handler
    (reitit/router routes))
   (reitit/create-resource-handler
    {:path "/"})
   (reitit/create-default-handler
    {:not-found
     (constantly (response/not-found "404 - Page not found"))
     :method-not-allowed
     (constantly (response/method-not-allowed "405 - Not allowed"))
     :not-acceptable
     (constantly (response/not-acceptable "406 - Not acceptable"))})))

(defn -main []
  (jetty/run-jetty
   (-> #'handler
       wrap-nocache
       wrap-reload)
   {:port 3000
    :join? false}))
